ee/tabby-webserver/src/service/mod.rs

```diff
@@ -129,3 +127,28 @@ impl ServerContext {
         }
     }
 }
+
+impl ServiceLocator for ServerContext {
+    fn auth(&self) -> &dyn AuthenticationService {
+        &self.db_conn
+    }
+
+    fn worker(&self) -> &dyn WorkerService {
+        self
+    }
+
+    fn code(&self) -> &dyn CodeSearch {
+        &*self.code
+    }
+
+    fn logger(&self) -> &dyn RawEventLogger {
+        &*self.logger
+    }
+}
+
+pub async fn create_service_locator(
```

Comment:
the only public method for entire service package is `service_locator`

cc @darknight